### PR TITLE
Integration tests for Precision in Lite

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Converts the precision plugin integration tests from PL to Lite tests where applicable. 
The tests can be classified into three groups (annotated in code for review):

- adopted: This test was fully or partially adopted into Lite
- skipped: The test is legacy or questionable whether it is of any use at all
- unimplemented: This test covers functionality only available in Trainer, and can't be ported to Lite. Example: Apex

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

cc @carmocca @justusschock @awaelchli @borda @akihironitta @rohitgr7

### CODE REVIEW

**Review by Borda — APPROVED** [review]

File: tests/tests_lite/helpers/models.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+import torch.nn as nn
+from torch import Tensor
+from torch.nn import Module
+from torch.optim import Optimizer
+from torch.utils.data import DataLoader, Dataset, IterableDataset
+
+from lightning_lite import LightningLite
+
+
+class RandomDataset(Dataset):
```
Comment (justusschock) [inline tests/tests_lite/helpers/models.py:13]: @carmocca @akihironitta @Borda is this something we might "offload" to the utils package so that we don't have a duplication of this in PL and Lite?

File: tests/tests_lite/plugins/precision/test_double_integration.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+class BoringDoubleModule(nn.Module):
+    def __init__(self):
+        super().__init__()
+        self.layer = torch.nn.Linear(32, 2)
+        self.register_buffer("complex_buffer", torch.complex(torch.rand(10), torch.rand(10)), False)
+
+    def forward(self, x):
+        assert x.dtype == torch.float64
+        assert torch.tensor([0.0]).dtype == torch.float64
+        assert torch.tensor([0.0], dtype=torch.float16).dtype == torch.float16
```
Comment (justusschock) [inline tests/tests_lite/plugins/precision/test_double_integration.py:30]: why this check?  If this doesn't hold this is an issue with torch not on our side.

File: tests/tests_lite/plugins/precision/test_double_integration.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        return self.layer(x)
+
+
+class DoublePrecisionBoringLite(BoringLite):
+    def get_model(self):
+        return BoringDoubleModule()
+
+    def step(self, model, batch):
+        model.double()  # TODO(lite): this needs to be done automatically in Lite.setup()
+        assert model.layer.weight.dtype == torch.float64
```
Comment (justusschock) [inline tests/tests_lite/plugins/precision/test_double_integration.py:40]:
```suggestion
        assert model.layer.weight.dtype == model.layer.bias.dtype == torch.float64
```

**Review by justusschock — APPROVED** [review]

File: tests/tests_lite/helpers/models.py (hunk 1 above)
Comment (carmocca) [reply tests/tests_lite/helpers/models.py:13]: No because this requires PyTorch to be installed.

**Review by carmocca — APPROVED** [review]
